cmd/talosctl/pkg/talos/action/node.go

```diff
@@ -0,0 +1,321 @@
+
+			a.update(reporter.Update{
+				Message: "Post check passed",
+				Status:  reporter.StatusSucceeded,
+			})
+
+			return nil
+		}()
+
+		// handle retryable errors
+
+		if errors.Is(err, context.Canceled) {
+			return nil
+		}
+
+		statusCode := client.StatusCode(err)
+		if statusCode == codes.Unavailable {
+			a.update(reporter.Update{
+				Message: "Unavailable, retrying...",
+				Status:  reporter.StatusError,
+			})
+
+			return retry.ExpectedError(err)
+		}
+
+		return err
+	})
+}
+
+func (a *nodeTracker) handleEvents(eventCh chan client.EventResult, actorID string) error {
+	for {
+		var eventResult client.EventResult
+
+		select {
+		case eventResult = <-eventCh:
+		case <-a.ctx.Done():
+			return a.ctx.Err()
+		}
+
+		if a.tracker.expectedEventFn(eventResult) {
+			status := reporter.StatusSucceeded
+			if a.tracker.postCheckFn != nil {
+				status = reporter.StatusRunning
+			}
+
+			a.update(reporter.Update{
+				Message: "Events check condition met",
+				Status:  status,
+			})
+
+			return nil
+		}
+
+		if eventResult.Error != nil {
+			a.update(reporter.Update{
+				Message: fmt.Sprintf("Error: %v", eventResult.Error),
+				Status:  reporter.StatusError,
+			})
+
+			return eventResult.Error
+		}
+
+		if eventResult.Event.ActorID == actorID {
+			err := a.handleEvent(eventResult.Event)
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func (a *nodeTracker) handleEvent(event client.Event) error {
+	switch msg := event.Payload.(type) {
+	case *machineapi.PhaseEvent:
+		a.update(reporter.Update{
+			Message: fmt.Sprintf("PhaseEvent phase: %s action: %v", msg.GetPhase(), msg.GetAction()),
+			Status:  reporter.StatusRunning,
+		})
+
+	case *machineapi.TaskEvent:
+		a.update(reporter.Update{
+			Message: fmt.Sprintf("TaskEvent task: %s action: %v", msg.GetTask(), msg.GetAction()),
+			Status:  reporter.StatusRunning,
+		})
+
+		if msg.GetTask() == "stopAllServices" {
+			return retry.ExpectedError(fmt.Errorf("stopAllServices task completed"))
+		}
+
+	case *machineapi.SequenceEvent:
+		errStr := ""
+		if msg.GetError().GetMessage() != "" {
+			errStr = fmt.Sprintf(
+				" error: [code: %v message: %v]",
+				msg.GetError().GetMessage(),
+				msg.GetError().GetCode(),
+			)
+		}
+
+		a.update(reporter.Update{
+			Message: fmt.Sprintf("SequenceEvent sequence: %s action: %v%v", msg.GetSequence(), msg.GetAction(), errStr),
+			Status:  reporter.StatusRunning,
+		})
+
+	case *machineapi.MachineStatusEvent:
+		a.update(reporter.Update{
+			Message: fmt.Sprintf("MachineStatusEvent: stage: %v ready: %v unmetCond: %v", msg.GetStage(), msg.GetStatus().GetReady(), msg.GetStatus().GetUnmetConditions()),
+			Status:  reporter.StatusRunning,
+		})
+
+	case *machineapi.ServiceStateEvent:
+		a.update(reporter.Update{
+			Message: fmt.Sprintf("ServiceStateEvent service: %v message: %v healthy: %v", msg.GetService(), msg.GetMessage(), msg.GetHealth().GetHealthy()),
+			Status:  reporter.StatusRunning,
+		})
+
+	case *machineapi.AddressEvent:
```

Comment:
this is more of a hacky event which we used to broadcast machine addresses, it gets sent periodically, idk, probably we should ignore it?

Reply:
done